docs/rfc/rfc-016-node-architecture.md

```diff
@@ -0,0 +1,83 @@
+
+The decisions made at the node level have the biggest impact to simplifying the protocols within them, unlocking better internal designs and making Tendermint more intuitive to use and easier to understand from the outside. Work, in minor increments, has already begun on this section of the codebase. This document exists to spark forth the necessary discourse in a few related areas that will help the team to converge on the long term makeup of the node.
+
+## Discussion
+
+The following is a list of points of discussion around the architecture of the node:
+
+### Dependency Tree
+
+The node object is currently stuffed with every component that possibly exists within Tendermint. In the constructor, all objects are built and interlaid with one another in some awkward dance. My guiding principle is that the node should only be made up of the components that it wants to have direct control of throughout its life. The node is a service which currently has the purpose of starting other services up in a particular order and stopping them all when commanded to do so. However, there are many services which are not direct dependents i.e. the mempool and evidence services should only be working when the consensus service is running. I propose to form more of a hierarchical structure of dependents which forces us to be clear about the relations that one component has to the other. More concretely, I propose the following dependency tree:
+
+![node dependency tree](./images/node-dependency-tree.svg)
+
+Many of the further discussion topics circle back to this representation of the node.
+
+As pointed out by Michael (@creachadair), It's also important to distinguish two dimensions which may require different characteristics of the architecture. There is the starting and stopping of services and their general lifecycle management. What is the correct oder of operations to starting a node for example. Then there is the question of the needs of the service during actual operation. Does it require access to events? Which data does it need read access from and so forth.
+
+An alternative model and one that perhaps better suits the latter of these dimensions is the notion of an internal message passing system. Either the events bus or p2p layer could serve as a viable transport. This would essentially allow all services to communicate with any other service and could perhaps provide a solution to the coordination problem (presented below) without a centralized coordinator. The other main advantage is that such a system would be more robust to disruptions and changes to the code which may make a hierarchical structure quickly outdated and suboptimal. The addition of message routing is an added complexity to implement, will increase the degree of asynchronicity in the system and may make it harder to debug problems that are across multiple services.
```

Comment:
I have a comment at the end of this in regards to putting each communication protocol into a category of how it's used and who it's used for.

Reply:
To your question, does wanting to have multiple nodes run in a single process really make a difference to whether we use p2p or the event bus for internal cases? It seems quite orthogonal to me

Reply:
> To your question, does wanting to have multiple nodes run in a single process really make a difference to whether we use p2p or the event bus for internal cases? It seems quite orthogonal to me

I mean, it doesn't necessarily make a difference. But if we were to use p2p for internode communication within the process, someone is going to complain that the loopback is less efficient than direct function calls, and ask for an optimization. And at that point, the line between the two kind of blurs.

On the other hand, if we think loopback transport is fast enough for internode within a single process, it could be simpler to use it for intranode too. Having said that, I don't think this is something we have to decide ahead of time. Your point about classifying by use still holds—it's just that we might use the same solution for different uses.

Reply:
This is a kind of optimization (direct function calls vs loopback) that doesn't pay off, as it make the design more complex and redundant, while possibly providing a minimal performance improvement. I am in line with what I think it is the spirit of this RFC, which is to make communication more abstract, which in this case means opting for the loopback approach.